cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,803 @@
+    return construct_quadtree<T>(x, y, static_cast<T>(x1), static_cast<T>(y1),
+                                 static_cast<T>(x2), static_cast<T>(y2),
+                                 static_cast<T>(scale), num_level, min_size, mr,
+                                 stream);
+  }
+
+  template <typename T,
+            std::enable_if_t<!std::is_floating_point<T>::value> * = nullptr>
+  inline std::unique_ptr<cudf::experimental::table> operator()(
+      cudf::mutable_column_view &x, cudf::mutable_column_view &y,
+      double const x1, double const y1, double const x2, double const y2,
+      double const scale, cudf::size_type const num_level,
+      cudf::size_type const min_size, rmm::mr::device_memory_resource *mr,
+      cudaStream_t stream) {
+    CUDF_FAIL("Non-floating point operation is not supported");
+  }
+};
+
+}  // namespace
+
+namespace cuspatial {
+
+namespace detail {
+
+std::unique_ptr<cudf::experimental::table> quadtree_on_points(
+    cudf::mutable_column_view x, cudf::mutable_column_view y, double const x1,
+    double const y1, double const x2, double const y2, double const scale,
+    cudf::size_type const num_levels, cudf::size_type const min_size,
+    rmm::mr::device_memory_resource *mr, cudaStream_t stream) {
+  return cudf::experimental::type_dispatcher(
+      x.type(), dispatch_construct_quadtree{}, x, y, x1, y1, x2, y2, scale,
+      num_levels, min_size, mr, stream);
+}
+
+}  // namespace detail
+
+std::unique_ptr<cudf::experimental::table> quadtree_on_points(
+    cudf::mutable_column_view x, cudf::mutable_column_view y, double const x1,
+    double const y1, double const x2, double const y2, double const scale,
+    cudf::size_type const num_levels, cudf::size_type const min_size,
+    rmm::mr::device_memory_resource *mr) {
+  CUSPATIAL_EXPECTS(x.size() == y.size(),
+                    "x and y columns might have the same length");
+  CUSPATIAL_EXPECTS(x1 < x2 && y1 < y2, "invalid bounding box (x1,y1,x2,y2)");
+  CUSPATIAL_EXPECTS(scale > 0, "scale must be positive");
+  CUSPATIAL_EXPECTS(num_levels >= 0 && num_levels < 16,
+                    "maximum of levels might be in [0,16)");
+  CUSPATIAL_EXPECTS(
+      min_size > 0,
+      "minimum number of points for a non-leaf node must be larger than zero");
+  if (x.is_empty() || y.is_empty()) {
+    std::vector<std::unique_ptr<cudf::column>> cols{};
+    cols.reserve(5);
+    cols.push_back(make_fixed_width_column<int32_t>(0, 0, mr));
+    cols.push_back(make_fixed_width_column<int8_t>(0, 0, mr));
+    cols.push_back(make_fixed_width_column<bool>(0, 0, mr));
+    cols.push_back(make_fixed_width_column<int32_t>(0, 0, mr));
+    cols.push_back(make_fixed_width_column<int32_t>(0, 0, mr));
+    return std::make_unique<cudf::experimental::table>(std::move(cols));
+  }
+  return detail::quadtree_on_points(x, y, x1, y1, x2, y2, scale, num_levels,
+                                    min_size, mr, cudaStream_t{0});
+}
+
+}  // namespace cuspatial
```

Comment:
800 line file... consider splitting?

Reply:
First attempt here: https://github.com/rapidsai/cuspatial/pull/143/commits/54242dd5034fd4011146b8ea93308548628bf709

This commit doesn't make it any less code, but it separates the functions related to phase 1 (full non-empty subquadrant construction), and phase 2 (pruning/merging quadrants with fewer than `min_size` points). Thoughts?